### PR TITLE
Sonar support phase 1

### PR DESCRIPTION
**Pull request by digitalentity**
- Re-enable sonar support
- Proper hardware detection for HC-SR04 hardware
- Enable sonar for QUANTON (M7 & M8)

Terrain following (`SURFACE` mode is not ready yet)!

Configurator support for rangefinder hardware is required.

### CODE REVIEW

**Comment by digitalentity** [comment]
I suggest we merge this set of changes - It's fairly local.

File: src/main/drivers/sonar_hcsr04.c [hunk 1]
```diff
@@ -86,79 +96,136 @@ void hcsr04_init(void)
...
+             * 340 m/s = 0.034 cm/microsecond = 29.41176471 *2 = 58.82352941 rounded to 59 
+             */
+
+            lastCalculatedDistance = hcsr04SonarPulseTravelTime / 59;
+            if (lastCalculatedDistance > HCSR04_MAX_RANGE_CM) {
+                lastCalculatedDistance = RANGEFINDER_OUT_OF_RANGE;
+            }
+        }
+        else {
+            /* No measurement within reasonable time - indicate failure */
```
Comment (martinbudden) [inline src/main/drivers/sonar_hcsr04.c]: I prefer `//` for in-code comments, and to reserve `/*` for comments at the top of a function.

**Comment by martinbudden** [comment]
Agreed, it's fairly local, so let's merge it in.

**Comment by digitalentity** [comment]
Will change comments and merge

**Comment by stronnag** [comment]
Bench testing seems much more reliable than the old implementation (HCSR04).

**Comment by stronnag** [comment]
For what it's worth, I flew this today and two logs are available if useful:

http://seyrsnys.myzen.co.uk/inav_1.6.2_sonar/LOG00626_1.TXT : PosHold at c. 2m altitude prior to mission (25m). RTH/Land; so there will be some useful sonarRaw at the start and end of the log.

http://seyrsnys.myzen.co.uk/inav_1.6.2_sonar/LOG00627_1.TXT : Mission at 5m (damn, I'd mis-remembered 3m). PosHold at c.1m for about the last 1 minute, so there should be some real sonarRaw at the end.

In both cases over soft ground, wet grass.

Question: What is sonarRaw? A time (in what units)?, an altitude (in what units)?

**Comment by digitalentity** [comment]
@stronnag thanks!

That sonar of yours is picking up too much noise. Seems like it's useless (reads garbage) for altitudes above 30-35cm. To cut that noise  you may want to `set inav_max_sonar_altitude = 30`.

`sonarRaw` is altitude in [cm] as reported by sonar sensor.

**Comment by stronnag** [comment]
Sensor is HCSR04. Mounted on double-sided sticky foam pads (and then a zip tie). In the house, over 'hard' floors it seems reliable over 30cm (I'll do some tests later and update this post). 

The flight surface was very soft, wet grass and uneven ground.

So I walked the machine around the house, up the stairs, held over the banister etc. Over worn out carpet it seems to be reasonable to perhaps 1m.
sonar
[sonar.txt](https://github.com/iNavFlight/inav/files/839513/sonar.txt). File is tab delimited, -1 for no data.

may be I have another HCSR04 device in a box somewhere.

**Pull request merged**